File: 2025/day11/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Math::BigInt;

# Parse input into a graph (hash of arrays)
sub parse_input {
    my ($filename) = @_;
    my %graph;

    open(my $fh, '<', $filename) or die "Cannot open $filename: $!";
    while (my $line = <$fh>) {
        chomp $line;
        next if $line =~ /^\s*$/;

        my ($node, $neighbors_str) = split(/: /, $line, 2);
        my @neighbors = split(/\s+/, $neighbors_str || '');
        $graph{$node} = \@neighbors;
    }
    close($fh);

    return \%graph;
}

# Part 1: Count all paths from 'you' to 'out' using memoization
sub part1 {
    my ($graph) = @_;
    my %memo;

    sub count_paths {
        my ($node, $graph, $memo) = @_;

        return $memo->{$node} if exists $memo->{$node};

        if ($node eq 'out') {
            $memo->{$node} = 1;
            return 1;
        }

        if (!exists $graph->{$node}) {
            $memo->{$node} = 0;
            return 0;
        }

        my $total = 0;
        for my $neighbor (@{$graph->{$node}}) {
            $total += count_paths($neighbor, $graph, $memo);
        }

        $memo->{$node} = $total;
        return $total;
    }

    return count_paths('you', $graph, \%memo);
}

# Helper function to count paths to a specific target
sub count_paths_to_target {
    my ($start, $target, $graph) = @_;
    my %memo;

    my $count_recursive;
    $count_recursive = sub {
        my ($node) = @_;

        return $memo{$node} if exists $memo{$node};

        if ($node eq $target) {
            $memo{$node} = Math::BigInt->new(1);
            return $memo{$node};
        }

        if (!exists $graph->{$node}) {
            $memo{$node} = Math::BigInt->new(0);
            return $memo{$node};
        }

        my $total = Math::BigInt->new(0);
        for my $neighbor (@{$graph->{$node}}) {
            $total += $count_recursive->($neighbor);
        }

        $memo{$node} = $total;
        return $total;
    };

    return $count_recursive->($start);
}

# Part 2: Count paths from 'svr' to 'out' that visit both 'dac' and 'fft'
sub part2 {
    my ($graph) = @_;

    # Count paths for each segment
    my $svr_to_dac = count_paths_to_target('svr', 'dac', $graph);
    my $svr_to_fft = count_paths_to_target('svr', 'fft', $graph);
    my $dac_to_fft = count_paths_to_target('dac', 'fft', $graph);
    my $fft_to_dac = count_paths_to_target('fft', 'dac', $graph);
    my $dac_to_out = count_paths_to_target('dac', 'out', $graph);
    my $fft_to_out = count_paths_to_target('fft', 'out', $graph);

    # Paths that visit dac before fft: svr -> dac -> fft -> out
    my $dac_before_fft = $svr_to_dac * $dac_to_fft * $fft_to_out;

    # Paths that visit fft before dac: svr -> fft -> dac -> out
    my $fft_before_dac = $svr_to_fft * $fft_to_dac * $dac_to_out;

    return $dac_before_fft + $fft_before_dac;
}

# Main
my $input_file = '../input.txt';
if (@ARGV > 0) {
    $input_file = $ARGV[0];
}

my $graph = parse_input($input_file);

print "Part 1: ", part1($graph), "\n";
print "Part 2: ", part2($graph), "\n";

File: 2025/day11/php/solution.php
#!/usr/bin/env php
<?php

function parse_input(string $filename): array {
    $graph = [];
    $lines = file($filename, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);

    foreach ($lines as $line) {
        $parts = explode(': ', $line, 2);
        $node = $parts[0];
        $neighbors = isset($parts[1]) ? explode(' ', $parts[1]) : [];
        $graph[$node] = $neighbors;
    }

    return $graph;
}

function part1(array $graph): int {
    $counter = new PathCounter($graph);
    return $counter->count_paths('you', 'out');
}

// Helper class for memoized path counting
class PathCounter {
    private array $graph;
    private array $memo_cache = [];

    public function __construct(array $graph) {
        $this->graph = $graph;
    }

    private function count_helper(string $node, string $target, array &$memo): int {
        if ($node === $target) {
            return 1;
        }

        if (isset($memo[$node])) {
            return $memo[$node];
        }

        if (!isset($this->graph[$node])) {
            $memo[$node] = 0;
            return 0;
        }

        $total = 0;
        foreach ($this->graph[$node] as $neighbor) {
            $total += $this->count_helper($neighbor, $target, $memo);
        }

        $memo[$node] = $total;
        return $total;
    }

    public function count_paths(string $start_node, string $target): int {
        if (!isset($this->memo_cache[$target])) {
            $this->memo_cache[$target] = [];
        }

        return $this->count_helper($start_node, $target, $this->memo_cache[$target]);
    }
}

function part2(array $graph): string {
    $counter = new PathCounter($graph);

    // Paths that visit dac before fft: svr -> dac -> fft -> out
    $svr_to_dac = $counter->count_paths('svr', 'dac');
    $dac_to_fft = $counter->count_paths('dac', 'fft');
    $fft_to_out = $counter->count_paths('fft', 'out');

    // Use bcmul for large number multiplication to avoid overflow
    $dac_before_fft = bcmul(bcmul((string)$svr_to_dac, (string)$dac_to_fft), (string)$fft_to_out);

    // Paths that visit fft before dac: svr -> fft -> dac -> out
    $svr_to_fft = $counter->count_paths('svr', 'fft');
    $fft_to_dac = $counter->count_paths('fft', 'dac');
    $dac_to_out = $counter->count_paths('dac', 'out');

    $fft_before_dac = bcmul(bcmul((string)$svr_to_fft, (string)$fft_to_dac), (string)$dac_to_out);

    return bcadd($dac_before_fft, $fft_before_dac);
}

function main(): void {
    global $argc, $argv;

    $input_file = '../input.txt';
    if ($argc > 1) {
        $input_file = $argv[1];
    }

    $graph = parse_input($input_file);

    echo "Part 1: " . part1($graph) . "\n";
    echo "Part 2: " . part2($graph) . "\n";
}

main();
